core/src/test/java/org/apache/iceberg/TestMetrics.java

```diff
@@ -96,67 +97,54 @@
       required(13, "timestampColBelowEpoch", TimestampType.withoutZone())
   );
 
+  private static final Schema FLOAT_DOUBLE_ONLY_SCHEMA = new Schema(
+      optional(1, "floatCol", FloatType.get()),
+      optional(2, "doubleCol", DoubleType.get())
+  );
+
   private final byte[] fixed = "abcd".getBytes(StandardCharsets.UTF_8);
 
   public abstract FileFormat fileFormat();
 
-  public Metrics getMetrics(InputFile file) {
-    return getMetrics(file, MetricsConfig.getDefault());
-  }
-
-  public abstract Metrics getMetrics(InputFile file, MetricsConfig metricsConfig);
+  public abstract Metrics getMetrics(Schema schema, MetricsConfig metricsConfig, Record... records) throws IOException;
 
-  public abstract InputFile writeRecords(Schema schema, Record... records) throws IOException;
+  public abstract Metrics getMetrics(Schema schema, Record... records) throws IOException;
```

Comment:
This refactor seems to have introduced a lot of changes. Is it needed? Seems like it may just introduce conflicts.

Reply:
I think the main reason for refactoring is that, before this change we have `writeRecords` to create appender and return `InputFile`, and then in Parquet specific tests we use `ParquetUtil.fileMetrics` to read metrics from the file footer of `InputFile` directly. But now since NaN is tracked during writing, to test NaN we will need to test against `appender.metrics()`.

Reply:
Makes sense! Looks like we need to keep it then.